File: src/lib/ticket-utils.js
const {
  ChannelType,
  PermissionFlagsBits,
  EmbedBuilder,
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  AttachmentBuilder
} = require('discord.js');

function formatTicketName(prefix, counter) {
  return `${prefix}-${String(counter).padStart(3, '0')}`;
}

function panelComponents() {
  return [
    new ActionRowBuilder().addComponents(
      new ButtonBuilder()
        .setCustomId('ticket_open')
        .setLabel('Open Ticket')
        .setEmoji('🎟️')
        .setStyle(ButtonStyle.Primary)
    )
  ];
}

function controlPanelComponents() {
  return [
    new ActionRowBuilder().addComponents(
      new ButtonBuilder().setCustomId('ticket_claim').setLabel('Claim').setEmoji('👤').setStyle(ButtonStyle.Secondary),
      new ButtonBuilder().setCustomId('ticket_close').setLabel('Close').setEmoji('🔒').setStyle(ButtonStyle.Danger),
      new ButtonBuilder().setCustomId('ticket_reopen').setLabel('Reopen').setEmoji('🔁').setStyle(ButtonStyle.Success),
      new ButtonBuilder().setCustomId('ticket_transcript').setLabel('Transcript').setEmoji('📁').setStyle(ButtonStyle.Primary),
      new ButtonBuilder().setCustomId('ticket_delete').setLabel('Delete').setEmoji('🗑️').setStyle(ButtonStyle.Danger)
    )
  ];
}

function sanitizeChannelName(name) {
  const safe = name
    .toLowerCase()
    .replace(/[^a-z0-9-_]/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '')
    .slice(0, 90);

  return safe || `ticket-${Date.now().toString().slice(-6)}`;
}

function escapeHtml(value) {
  return value
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}

async function createTicketChannel(guild, opener, ticketNumber, setup) {
  const channelName = formatTicketName(setup.ticketPrefix || 'ticket', ticketNumber);
  const overwrites = [
    {
      id: guild.roles.everyone.id,
      deny: [PermissionFlagsBits.ViewChannel]
    },
    {
      id: opener.id,
      allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.ReadMessageHistory]
    }
  ];

  for (const roleId of setup.supportRoleIds) {
    overwrites.push({
      id: roleId,
      allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages, PermissionFlagsBits.ReadMessageHistory, PermissionFlagsBits.ManageChannels]
    });
  }

  const channel = await guild.channels.create({
    name: channelName,
    type: ChannelType.GuildText,
    parent: setup.ticketCategoryId,
    permissionOverwrites: overwrites,
    topic: `Ticket #${ticketNumber} | Owner: ${opener.id}`
  });

  await channel.send({
    content: `${opener} أهلاً بك، تم فتح تذكرتك بنجاح.`,
    embeds: [
      new EmbedBuilder()
        .setColor(0x5865f2)
        .setTitle(`Ticket #${String(ticketNumber).padStart(3, '0')}`)
        .setDescription('فريق الدعم سيرد عليك قريباً. استخدم الأزرار للإدارة.')
        .setTimestamp()
    ],
    components: controlPanelComponents()
  });

  return channel;
}

async function createTicketTranscript(channel, ticketNumber) {
  const collected = [];
  let before;

  for (let i = 0; i < 10; i += 1) {
    const fetched = await channel.messages.fetch({ limit: 100, before });
    if (fetched.size === 0) break;

    const batch = [...fetched.values()];
    collected.push(...batch);
    before = batch[batch.length - 1].id;

    if (fetched.size < 100) break;
  }

  const ordered = collected.sort((a, b) => a.createdTimestamp - b.createdTimestamp);
  const rows = ordered
    .map((message) => {
      const time = new Date(message.createdTimestamp).toISOString();
      const author = escapeHtml(`${message.author.tag} (${message.author.id})`);
      const content = escapeHtml(message.content || '[Attachment/Embed]');
      return `<tr><td>${time}</td><td>${author}</td><td>${content}</td></tr>`;
    })
    .join('\n');

  const html = `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<title>Ticket #${String(ticketNumber).padStart(3, '0')}</title>
<style>
body { font-family: Arial, sans-serif; background: #111; color: #f1f1f1; }
h1 { color: #8ab4ff; }
table { width: 100%; border-collapse: collapse; }
th, td { border: 1px solid #333; padding: 8px; text-align: left; vertical-align: top; }
th { background: #1f1f1f; }
tr:nth-child(even) { background: #171717; }
</style>
</head>
<body>
<h1>Transcript - Ticket #${String(ticketNumber).padStart(3, '0')}</h1>
<table>
<thead>
<tr><th>Time</th><th>User</th><th>Message</th></tr>
</thead>
<tbody>
${rows}
</tbody>
</table>
</body>
</html>`;

  return new AttachmentBuilder(Buffer.from(html, 'utf8'), {
    name: `ticket-${String(ticketNumber).padStart(3, '0')}.html`
  });
}

function buildLogEmbed({ title, description, user, moderator, channel, ticketNumber, guildName }) {
  return new EmbedBuilder()
    .setColor(0x2b2d31)
    .setTitle(title)
    .setDescription(description)
    .addFields(
      { name: 'User', value: user ? `${user.tag} (${user.id})` : 'N/A', inline: true },
      { name: 'Moderator', value: moderator ? `${moderator.tag} (${moderator.id})` : 'N/A', inline: true },
      { name: 'Channel', value: channel ? `${channel}` : 'N/A', inline: true },
      { name: 'Ticket Number', value: `#${String(ticketNumber).padStart(3, '0')}`, inline: true },
      { name: 'Server', value: guildName || 'Unknown', inline: true },
      { name: 'Time', value: `<t:${Math.floor(Date.now() / 1000)}:F>`, inline: false }
    )
    .setTimestamp();
}

module.exports = {
  panelComponents,
  sanitizeChannelName,
  createTicketChannel,
  createTicketTranscript,
  buildLogEmbed,
  formatTicketName
};

File: src/lib/ticket_utils.py
from __future__ import annotations

import html
from datetime import datetime, timezone
from io import BytesIO
from typing import Any, Dict

import discord


def format_ticket_name(prefix: str, counter: int) -> str:
    return f"{prefix}-{counter:03d}"


def sanitize_channel_name(name: str) -> str:
    sanitized = "".join(ch.lower() if ch.isalnum() or ch in "-_" else "-" for ch in name)
    while "--" in sanitized:
        sanitized = sanitized.replace("--", "-")
    sanitized = sanitized.strip("-")[:90]
    return sanitized or f"ticket-{datetime.now(timezone.utc).strftime('%H%M%S')}"


def build_panel_view() -> discord.ui.View:
    view = discord.ui.View(timeout=None)
    view.add_item(
        discord.ui.Button(
            custom_id="ticket_open",
            label="Open Ticket",
            emoji="🎟️",
            style=discord.ButtonStyle.primary,
        )
    )
    return view


def build_control_view() -> discord.ui.View:
    view = discord.ui.View(timeout=None)
    controls = [
        ("ticket_claim", "Claim", "👤", discord.ButtonStyle.secondary),
        ("ticket_close", "Close", "🔒", discord.ButtonStyle.danger),
        ("ticket_reopen", "Reopen", "🔁", discord.ButtonStyle.success),
        ("ticket_transcript", "Transcript", "📁", discord.ButtonStyle.primary),
        ("ticket_delete", "Delete", "🗑️", discord.ButtonStyle.danger),
    ]
    for custom_id, label, emoji, style in controls:
        view.add_item(discord.ui.Button(custom_id=custom_id, label=label, emoji=emoji, style=style))
    return view


def build_log_embed(payload: Dict[str, Any]) -> discord.Embed:
    ticket_number = int(payload.get("ticket_number", 0))
    embed = discord.Embed(
        title=payload.get("title", "Ticket Log"),
        description=payload.get("description", ""),
        color=0x2B2D31,
        timestamp=datetime.now(timezone.utc),
    )
    user = payload.get("user")
    moderator = payload.get("moderator")
    channel = payload.get("channel")
    guild_name = payload.get("guild_name", "Unknown")

    embed.add_field(name="User", value=f"{user} ({getattr(user, 'id', 'N/A')})" if user else "N/A", inline=True)
    embed.add_field(name="Moderator", value=f"{moderator} ({getattr(moderator, 'id', 'N/A')})" if moderator else "N/A", inline=True)
    embed.add_field(name="Channel", value=f"{channel.mention if channel else 'N/A'}", inline=True)
    embed.add_field(name="Ticket Number", value=f"#{ticket_number:03d}", inline=True)
    embed.add_field(name="Server", value=guild_name, inline=True)
    return embed


async def create_ticket_transcript(channel: discord.TextChannel, ticket_number: int) -> discord.File:
    messages = []
    async for message in channel.history(limit=1000, oldest_first=True):
        messages.append(message)

    rows = []
    for message in messages:
        stamp = message.created_at.replace(tzinfo=timezone.utc).isoformat()
        author = html.escape(f"{message.author} ({message.author.id})")
        content = html.escape(message.content or "[Attachment/Embed]")
        rows.append(f"<tr><td>{stamp}</td><td>{author}</td><td>{content}</td></tr>")

    document = f"""<!DOCTYPE html>
<html lang='en'>
<head>
  <meta charset='utf-8'>
  <title>Ticket #{ticket_number:03d}</title>
  <style>
    body {{ background: #111; color: #f3f3f3; font-family: Arial, sans-serif; }}
    table {{ width: 100%; border-collapse: collapse; }}
    th, td {{ border: 1px solid #333; padding: 8px; text-align: left; vertical-align: top; }}
  </style>
</head>
<body>
  <h1>Transcript - Ticket #{ticket_number:03d}</h1>
  <table>
    <thead><tr><th>Time</th><th>User</th><th>Message</th></tr></thead>
    <tbody>{''.join(rows)}</tbody>
  </table>
</body>
</html>"""

    return discord.File(BytesIO(document.encode("utf-8")), filename=f"ticket-{ticket_number:03d}.html")
